File: src/ch/waan/mcddpui/structure/Record.scala
package ch.waan.mcddpui.structure

import scala.util.Try
import scala.util.Failure
import scala.util.Success

trait Record[T] {
  def update(f: T => Try[T]): Option[Throwable]
  def passTo(f: T => Option[Throwable]): Option[Throwable]
  def undo: Option[RecordHistoryManipulationException]
  def redo: Option[RecordHistoryManipulationException]
}

class RecordHistoryManipulationException(msg: String) extends Exception(msg)

File: src/ch/waan/mcddpui/api/Record.java
package ch.waan.mcddpui.api;

public interface Record<T> {

	void view(ReadCommand<? super T> command) throws Throwable;

	void update(MutationCommand<? super T, ? extends T> command) throws Throwable;

	void undo() throws RecordHistoryManipulationException;

	void redo() throws RecordHistoryManipulationException;

}